### ISSUE TITLE
Enabling HTML translation gives truncated Output

### ISSUE THREAD

**Issue by abhi-agg** [browsermt/bergamot-translator#273]
I enabled the HTML translation for the `native` mode of the bergamot translator app (by setting `ResponseOptions::alignment = true` and `ResponseOptions::HTML = true`) and performed the inference. I noticed truncated translation on the terminal.

**Steps to reproduce**:

**Code**: main branch of this repo ([8e79897f30a3948621e95b657fda2bfc6f69bc76](https://github.com/browsermt/bergamot-translator/tree/8e79897f30a3948621e95b657fda2bfc6f69bc76)) + apply [this patch to enable HTML translation](https://github.com/abhi-agg/bergamot-translator/commit/c47f84dfd84b17ae27e69dd401788991aa418f4f)

**Platform**: Intel / MacOS Big Sur (11.6)

**Translator config used**:
```
relative-paths: true
models:
  - model.intgemm.alphas.bin
vocabs:
  - vocab.deen.spm
  - vocab.deen.spm
shortlist:
    - lex.s2t.bin
    - false
beam-size: 1
normalize: 1.0
word-penalty: 0
workspace: 2000
max-length-factor: 2.5
gemm-precision: int8shiftAlphaAll
max-length-break: 128
mini-batch-words: 1024
log-level: info
alignment: soft
```

where models and vocab files are downloaded fresh from `student` repository.

**Command used to run bergamot translator native binary in "native" mode**:

`echo "Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation, a non-profit organization that also hosts a range of other projects. Wikipedia is hosted by the Wikimedia Foundation. How are you doing my friend Wikimedia?" | ./app/bergamot --bergamot-mode native --cpu-threads 1  --model-config-paths $HOME/work/students/deen/ende.student.tiny11/config.intgemm8bitalpha.bergamot.yml`

**Output on terminal**:

```
[2021-12-13 15:14:06] [data] Loading SentencePiece vocabulary from file /Users/abhishek/work/students/deen/ende.student.tiny11/vocab.deen.spm
[2021-12-13 15:14:06] Missing list of protected prefixes for sentence splitting. Set with --ssplit-prefix-file.
[2021-12-13 15:14:06] [data] Loading binary shortlist as /Users/abhishek/work/students/deen/ende.student.tiny11/lex.s2t.bin false
[2021-12-13 15:14:06] [data] Lexical short list firstNum 50 and bestNum 50
[2021-12-13 15:14:06] [memory] Extending reserved space to 2048 MB (device cpu0)
[2021-12-13 15:14:06] Loaded model config
[2021-12-13 15:14:06] Loading scorer of type transformer as feature F0
[2021-12-13 15:14:06] [memory] Reserving 31 MB, device cpu0
[2021-12-13 15:14:06] [memory] Reserving 8 MB, device cpu0
```
`
Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation veranstaltet, einer gemeinnützigen Organisation, die auch eine Reihe anderer Projekte veranstaltet. Wikipedia wird von der Wikimedia Foundation gehostet. Wie geht es meinem Freund Wi%`

**Observation**:
1. Output (i.e. translated text) starts getting truncated from the last sentence `How are you doing my friend Wikimedia?`. If I remove the last word `Wikimedia` from the input text then everything is fine. Anything appended after the `How are you doing my friend?` is ignored in the output, indicating some sort of word/character/token limits coming into play.
2. If I let `ResponseOptions::alignment = true` but set `ResponseOptions::HTML = false` then everything is fine. => `ResponseOptions::HTML = true` is probably the reason.

cc @jelmervdl @jerinphilip @andrenatal

**Comment by jelmervdl** [comment]
Found the issue, it was due to fixed length buffers in xh_scanner.

I thought about just emitting TT_TEXT multiple times when the buffer was full, but noticed that in other places such an approach was less feasible, or whatever was there would still cause a data loss.

So I just replaced it all with std::string in #274.

**Comment by andrenatal** [comment]
Thanks for fixing this @jelmervdl.  I don't want to be pushy but this is blocking us heavily, since we took as granted that it was working by default after it was released to us and we already had time allotted to work on this. 

Would be possible to have it working before the holidays? Also, is it possible to expand the test coverage in order to cover the edge cases and increase the confidence in the functionality? It would be bad to start getting errors reports about broken pages from users after this is on the wild.

Thanks.

**Comment by jelmervdl** [comment]
This bug should have been caught by test cases. I would like to blame it on importing code without knowing all of its design decisions and limitations, but could have been easily caught by an integration test.

Re confidence, for me it would be really helpful to get an idea at the sort of complexity of HTML would be expected. I've looked at the extension, but currently it seems to [just submit text](https://github.com/mozilla/firefox-translations/blob/main/extension/view/js/InPageTranslation.js#L141) (even though [it says it is HTML](https://github.com/mozilla/firefox-translations/blob/main/extension/controller/translation/translationWorker.js#L417)).

I'm working to adding some integration tests to https://github.com/browsermt/bergamot-translator-tests with HTML that seems good (`innerHTML` of block-level elements from Wikipedia, Reddit, Twitter, some random recipe website) and bad (complete pages, not entirely valid HTML, etc.) just to see when it fails, how it fails. (@jerinphilip I might ask you about tips for this)

Maybe this should be documented as: 
- The HTML parsing and reconstruction is only compatible with fragments of valid HTML, by design.
- The parser cannot digest XML-isms like CDATA or self-closing tags,
- … and removes non-text but text-containing elements like `<style>` and `<script>`, or HTML comments.
- There is no guarantee on the order of the elements in the translated output. The hierarchy will stay intact (parent-child relations are guaranteed to be the same) but if you would pass it a full HTML page, it could move the `<body>` element before the `<head>` element, in theory.
- All elements that do return in the output will retain their attributes (in the same order)
- (not yet implemented) all *empty* elements occurring in the input HTML (i.e. `<img>` and `<div></div>`, but not `<div>␣</div>`) are guaranteed to be reproduced *at least once* in the output HTML.
- (problematic) The HTML in response.source might differ from the HTML you submitted, as this is reconstructed with the same method as the translated HTML is made. The smallest unit here is sentencepiece token, and HTML can only be inserted back between tokens. This makes it easy to guarantee that the byte offsets used for annotating the input and translation HTML line up. (I still doubt those offsets are of any use though when using `responseOptions.html = true`)

**Comment by andrenatal** [comment]
Thanks for the @jelmervdl . You don't need the extension in order to input html. Just browse to any page, capture a large chunk of html > 1000 chars for example, and try to translate it.

**Comment by jelmervdl** [comment]
It's not just long HTML pages. I've tested that, but unfortunately those pages where very HTML heavy and pretty light on text.

The effect of this is that text between tags that’s longer than 1024 characters was truncated. HTML tags around it, or HTML tags breaking it up, should not be lost. I.e. if you give it `<p>xxx<br><u>yyy</u></p>` it will still produce `<p>XXX<br><u>YYY</u></p>`  but  xxx and yyy were both truncated to their first 1024 characters before they were translated to XXX YYY.

**Comment by jelmervdl** [comment]
This was fixed (and there is a test for it now) in #274. See that pull request for more information.
